### PR TITLE
Configurable nested cross reference query limit

### PR DESCRIPTION
**Pull request by parkerduckworth**
### What's being changed:

The number of nested cross-references returned can now be configured with `QUERY_NESTED_CROSS_REFERENCE_LIMIT`. Note that a limit of <= 0 makes the number of nested crossrefs returned unlimited.

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/inverted/searcher_ref_filter.go [hunk 1]
```diff
@@ -245,3 +241,13 @@ func (r *refFilterExtractor) validate() error {
 
 	return nil
 }
+
+func determineRefLimit() int {
+	limit, err := strconv.ParseInt(os.Getenv("QUERY_REF_LIMIT"), 10, 64)
+	if err != nil {
+		limit = config.DefaultQueryReferenceLimit
+	} else if limit < 0 {
```
Comment (aliszka) [inline adapters/repos/db/inverted/searcher_ref_filter.go:249]: what if limit == 0? is it unlimited or 0?

File: adapters/repos/db/inverted/searcher_ref_filter.go [hunk 2]
```diff
@@ -245,3 +241,13 @@ func (r *refFilterExtractor) validate() error {
 
 	return nil
 }
+
+func determineRefLimit() int {
```
Comment (aliszka) [inline adapters/repos/db/inverted/searcher_ref_filter.go:245]: so basically by default limit is now set to `config.DefaultQueryReferenceLimit = 100k`, "unlimited" when env var explicitly set to negative value, or env var value when it is >= 0?

File: adapters/repos/db/inverted/searcher_ref_filter.go (hunk 1 above)
Comment (parkerduckworth) [reply adapters/repos/db/inverted/searcher_ref_filter.go:249]: If limit == 0, no refs will be returned. This was requested by @kcm but I am open to suggestions

File: adapters/repos/db/inverted/searcher_ref_filter.go (hunk 2 above)
Comment (parkerduckworth) [reply adapters/repos/db/inverted/searcher_ref_filter.go:245]: Currently, if env var is set to 0, no refs will be returned

File: adapters/repos/db/inverted/searcher_ref_filter.go (hunk 1 above)
Comment (aliszka) [reply adapters/repos/db/inverted/searcher_ref_filter.go:249]: thx, it is clear now. I would just add a comment that 0 is allowed and works exactly as 0 (nothing is returned)

**Review by aliszka — APPROVED** [review]

File: adapters/repos/db/inverted/searcher_ref_filter.go (hunk 1 above)
Comment (parkerduckworth) [reply adapters/repos/db/inverted/searcher_ref_filter.go:249]: ok, adding the comment 👍

**Review by aliszka — APPROVED** [review]

**Comment by parkerduckworth** [comment]
In practice, a limit of `0` will not work, due to the pagination total limit checks made in the db:
https://github.com/weaviate/weaviate/blob/05f9fb12c2cdb6e5ab341251d2a50648a32bd990/adapters/repos/db/search.go#L371

So, `0` will now indicate that an unlimited search is desired.

File: modules/reranker-cohere/clients/ranker_test.go [hunk 3]
```diff
@@ -205,10 +206,14 @@ func (f *testRankHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 			index = 3
 		}
+		f.responseLock.Lock()
 		f.response.Results = f.batchedResults[index]
+		f.responseLock.Unlock()
 	}
 
+	f.responseLock.Lock()
 	outBytes, err := json.Marshal(f.response)
+	f.responseLock.Unlock()
```
Comment (antas-marcin) [inline modules/reranker-cohere/clients/ranker_test.go:216]: hmm, I think that this change is not need as at the very top of this method there's a lock for whole method invocation, or is it?
```
f.lock.Lock()
defer f.lock.Unlock()
```

File: usecases/config/config_handler.go [hunk 4]
```diff
@@ -73,6 +73,7 @@ type Config struct {
 	Debug                               bool                     `json:"debug" yaml:"debug"`
 	QueryDefaults                       QueryDefaults            `json:"query_defaults" yaml:"query_defaults"`
 	QueryMaximumResults                 int64                    `json:"query_maximum_results" yaml:"query_maximum_results"`
+	QueryNestedRefLimit                 int64                    `json:"query_ref_limit" yaml:"query_ref_limit"`
```
Comment (antas-marcin) [inline usecases/config/config_handler.go:76]: let's make the json / yaml variable name the same as `QUERY_NESTED_CROSS_REFERENCE_LIMIT` environment variable name for example: `query_nested_cross_reference_limit`

File: modules/reranker-cohere/clients/ranker_test.go (hunk 3 above)
Comment (antas-marcin) [reply modules/reranker-cohere/clients/ranker_test.go:216]: but it can stay, it's just a test

Comment (parkerduckworth) [reply modules/reranker-cohere/clients/ranker_test.go:216]: Ah I had made this change before it hit master. I will rm and just rely on the already existing f.lock

File: usecases/config/config_handler.go (hunk 4 above)
Comment (parkerduckworth) [reply usecases/config/config_handler.go:76]: fixing

**Review by antas-marcin — DISMISSED** [review]

**Pull request merged**